src/Compatibility/Microsoft.DotNet.ApiCompatibility/ValidationSuppressions/SuppressionEngine.cs

```diff
@@ -0,0 +1,208 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Xml.Serialization;
+
+namespace Microsoft.DotNet.ValidationSuppression
+{
+    /// <summary>
+    /// Collection of Suppressions which is able to add suppressions, check if a specific error is suppressed, and write all suppressions
+    /// down to a file. The engine is thread-safe.
+    /// </summary>
+    public class SuppressionEngine
```

Comment:
Should the `SupressionEngine` also support `NoWarn`? Or the story for `NoWarn` should be handled individually by every consumer?

Reply:
The way it is used today it is not needed, because this is today used only by PackageLogger (and in the future by ApiCompat) both of which already take noWarn. For example, the Logger today will only be called with an error after it has validated that the error wasn't specified in NoWarn, so no need for the engine to check again. ApiCompat also takes NoWarn, so we could in the same way perform some filtering for NoWarn first, and then check if the suppression engine is suppressing the specific error.

Reply:
That way we would avoid having separate logic to filter, currently `NoWarn` is supported via `DiagnosticBag.Filter`

Reply:
Yeah, however with the suppression engine we would centralize all the suppression logic, so I think that's better. Rather than having multiple places where we filter diagnostics?

Reply:
That said, we can do that centralization later, just wanted to get your thoughts.

Reply:
sure, it will be easy to do this by just passing in the NoWarns to the Engine, and then the engine would just return `true` in `IsErrorSuppressed` if the error passed in has the DiagnosticId that matches one of the elements in NoWarn

Reply:
Will address on a separate PR.